### PR TITLE
Add TypeScript definitions

### LINKED ISSUE #127: Add TypeScript definition
If anyone wants to contribute, keep this in mind:

- It should adhere to https://github.com/sindresorhus/typescript-definition-style-guide
- You should have a lot of TypeScript experience.
- The pull request should be high-quality. Not sloppy.
- Docs can be copy-pasted from the readme.

### PR DESCRIPTION
**Pull request by jopemachine**
# Add typescript definitions

Thanks for the quick reply!

Here is the PR.

> Maybe do a pull request to add types to that package? Link to it here and I'll ping the maintainer to review it.

Ok, I will try to do a pull request to add types to cache-conf.

## Related

Fixes #127

* [cache-conf type definitions issue](https://github.com/SamVerschueren/cache-conf/issues/7)

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+interface ModifierKeyItem {
+	readonly valid?: boolean;
+	readonly title?: string;
+	readonly subtitle?: string;
+	readonly arg?: string;
+	readonly icon?: string;
+	readonly variables?: Record<string, string>;
+}
+
+interface ActionElement {
```
Comment (sindresorhus) [inline index.d.ts:70]: You need to export all these types.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+	```
+	*/
+	readonly quicklookurl?: string;
+
+	readonly variables?: Record<string, string>;
+}
+
+export default interface Alfy {
+	/**
+	@param {Record<string, ScriptFilterItem>} items
```
Comment (sindresorhus) [inline index.d.ts:215]: Missing description. Doesn't need type annotation.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+	//=> ['unicorn']
+	```
+	*/
+	matches: (input: string, list: (string | Record<string, unknown>)[], item?: string | ((item: string | Record<string, unknown>, input: string) => boolean)) => string[];
+
+	/**
+	@param {string[]} list
+	@param {string|((item: string | Record<string, unknown>, input: string) => boolean)} item
+	@returns {string[]}
+	@description Same as matches(), but with alfy.input as input
```
Comment (sindresorhus) [inline index.d.ts:251]:
```suggestion
	@description Same as matches(), but with `alfy.input` as input.
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Make sure you follow https://github.com/sindresorhus/typescript-definition-style-guide

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+	readonly icon?: string;
+	readonly variables?: Record<string, string>;
+}
+
+interface ActionElement {
+	readonly text?: string[];
+	readonly url?: string;
+	readonly file?: string;
+	readonly auto?: string;
+}
```
[the remark this answers is not in the document]
Comment (jopemachine) [reply index.d.ts:122]: I just added some comments about this.

I have never used `action` attributes, so I'm not sure if there my comments about this are all appropriate.

**Comment by jopemachine** [comment]
I'm sorry for missing several conventions guides and thank you for pointing it out.

May I request code review again to find out if there are points that I'm still missing?

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+	See alfred-config for more details.
+
+	@example
+	```
+	alfy.userConfig.get('apiKey');
+	//=> '16811cad1b8547478b3e53eae2e0f083'
+	```
+	*/
+	userConfig: Map<string, string>;
+}
```
Comment (LitoMore) [inline index.d.ts:460]: Newline required.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,460 @@
+import Conf from 'conf';
+import { GotOptions, GotUrl } from 'got';
+
+export interface FetchOptions extends GotOptions<string> {
+	/**
+	Parse response body with JSON.parse and set accept header to application/json.
+	@default true
```
Comment (sindresorhus) [inline index.d.ts:7]: Too cramped.
```suggestion

	@default true
```

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+	/**
+	User will get displaying large type with ⌘L.
+	*/
+	readonly largetype?: string;
+}
+
+/**
+	Defines what to change when the modifier key is pressed.
+	When you release the modifier key, it returns to the original item.
+	@interface ModifierKeyItem
```
Comment (sindresorhus) [inline index.d.ts:91]: This is moot.

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+	readonly copy?: string;
+
+	/**
+	User will get displaying large type with ⌘L.
+	*/
+	readonly largetype?: string;
+}
+
+/**
+	Defines what to change when the modifier key is pressed.
```
Comment (sindresorhus) [inline index.d.ts:89]: Should not have indentation.

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+	@example
+	```
+	Single Item:
+	"action": "Alfred is Great"
+
+	Multiple Items:
+	"action": ["Alfred is Great", "I use him all day long"]
+
+	For control over the content type of the action, you can use an object with typed keys as follows:
+	"action": {
```
Comment (sindresorhus) [inline index.d.ts:236]: Use a proper code block and `@example` tag. See the TS style guide.

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+	/**
+	Variables can be passed out of the script filter within a variables object.
+	*/
+	readonly variables?: Record<string, string>;
+}
+
+export default interface Alfy {
+	/**
+	Return output to Alfred.
+	@param items
```
Comment (sindresorhus) [inline index.d.ts:280]: Missing description.

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+	/**
+	Alfred metadata.
+	*/
+	alfred: {
+		version: string;
+		theme: string;
+		themeBackground: string;
+		themeSelectionBackground: string;
+		themeSubtext: string;
+		data: string;
```
Comment (sindresorhus) [inline index.d.ts:436]: These properties need doc comments.

File: index.d.ts [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+	```
+	console.log(alfy.icon.error);
+	//=> '/System/Library/CoreServices/CoreTypes.bundle/Contents/Resources/AlertStopIcon.icns'
+
+	console.log(alfy.icon.get('Clock'));
+	//=> '/System/Library/CoreServices/CoreTypes.bundle/Contents/Resources/Clock.icns'
+	```
+	*/
+	icon: {
+		get: (icon: string) => string;
```
Comment (sindresorhus) [inline index.d.ts:418]: Need doc comment.

File: index.d.ts [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+	/**
+	Whether the user currently has the workflow debugger open.
+	*/
+	debug: boolean
+
+	/**
+	Exports a Map with the user workflow configuration.
+	A workflow configuration allows your users to provide configuration information for the workflow. 
+	For instance, if you are developing a GitHub workflow, you could let your users provide their own API tokens.
+	See alfred-config for more details.
```
Comment (sindresorhus) [inline index.d.ts:451]: Should be linkified like the readme.

Make sure that all the text you copied from the readme retains its style, links, and inline code blocks.

File: index.d.ts [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+		data: string;
+		cache: string;
+		preferences: string;
+		preferencesLocalHash: string;
+	};
+
+	/**
+	Whether the user currently has the workflow debugger open.
+	*/
+	debug: boolean
```
Comment (sindresorhus) [inline index.d.ts:445]:
```suggestion
	debug: boolean;
```

File: index.d.ts [hunk 15]
```diff
@@ -0,0 +1,460 @@
...
+	```
+	*/
+	output: (items: ScriptFilterItem[], options?: OutputOptions) => void;
+
+	/**
+	Returns an string[] of items in list that case-insensitively contains input.
+	@param input
+	@param list
+	@param target
+	@returns string[] of items in list that case-insensitively contains input.
```
Comment (sindresorhus) [inline index.d.ts:302]: Don't mention the type. It's already clear from the actual types.

File: index.d.ts [hunk 16]
```diff
@@ -0,0 +1,460 @@
...
+
+	/**
+	By specifying "type": "file", this makes Alfred treat your result as a file on your system.
+	This allows the user to perform actions on the file like they can with Alfred's standard file filters.
+	When returning files, Alfred will check if the file exists before presenting that result to the user.
+	This has a very small performance implication but makes the results as predictable as possible.
+	If you would like Alfred to skip this check as you are certain that the files you are returning exist, you can use "type": "file:skipcheck".
+	@default "default"
+	*/
+	readonly type?: "default" | "file" | "file:skipcheck";
```
Comment (sindresorhus) [inline index.d.ts:215]: Use single quotes everywhere.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by LitoMore** [comment]
@jopemachine Would you mind upgrading xo to `0.39.1` in the pull request? The linter will help you fix all linter-related issues automatically.

**Comment by LitoMore** [comment]
@jopemachine You could go through the [Testing](https://github.com/sindresorhus/typescript-definition-style-guide#testing) to learn how to use [tsd](https://github.com/SamVerschueren/tsd) to write tests for your TypeScript type definitions.

**Comment by jopemachine** [comment]
@LitoMore, @sindresorhus I appreciate for detailed code reviews :)

I upgraded xo to `0.39.1` and fix some linting bugs (New bugs after upgrading `xo`)

And updated type declaration, added `tsd` test.

In tsd test, to test functions with return type void, I added `/* eslint-disable @typescript-eslint/no-confusing-void-expression */` comment on top.

If we don't need to test void returning functions, just let me know, I will remove the comment and tests.

**Comment by jopemachine** [comment]
It seems that [update-notifier](https://github.com/yeoman/update-notifier) in xo `0.39.1` uses `node >= 10`,

So test fails in Node 8.

May I ask how I should handle this?

**Comment by LitoMore** [comment]
> Test fails in Node 8.

@jopemachine Feel free to drop Node.js 8 from GitHub actions.

It seems the [got@9](https://github.com/sindresorhus/alfy/blob/main/.github/workflows/main.yml#L13) blocked us using Node.js 10 or higher. We could solve this problem later (in another PR) if you are interested.

**Comment by jopemachine** [comment]
> > Test fails in Node 8.
> 
> @jopemachine Feel free to drop Node.js 8 from GitHub Actions.
> 
> It seems the [got@9](https://github.com/sindresorhus/alfy/blob/main/.github/workflows/main.yml#L13) blocked us using Node.js 10 or higher. We could solve this problem later (in another PR) if you are interested.

Of course, I'm happy to have the opportunity to contribute to open source.

It seems that `got@12.x` is ESM module, should we upgrade got to `got@11.8.2`?

File: package.json [hunk 17]
```diff
@@ -10,13 +10,14 @@
 		"alfy-cleanup": "cleanup.js"
 	},
 	"engines": {
-		"node": ">=8"
+		"node": ">=10.x"
```
Comment (LitoMore) [inline package.json:13]:
```suggestion
		"node": ">=10"
```

**Comment by LitoMore** [comment]
>It seems that `got@12` is ESM, should we upgrade got to `got@11.8.2`?

Yes. Let's discuss this in the future pull request.

**Comment by jopemachine** [comment]
> > It seems that `got@12` is ESM, should we upgrade got to `got@11.8.2`?
> 
> @jopemachine Yes. Let's discuss this in the future pull request.

@LitoMore I made a PR about this issue [here](https://github.com/sindresorhus/alfy/pull/140).

File: index.test-d.ts [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+expectType<string[]>(alfy.inputMatches(['foo', 'unicorn']));
+
+expectType<ScriptFilterItem[]>(alfy.inputMatches(mockItems, 'title'));
+
+expectType<void>(alfy.error(new Error('some error')));
+
+expectType<void>(alfy.log('some message'));
+
+expectType<Promise<any>>(alfy.fetch('https://api.foo.com', {
```
Comment (LitoMore) [inline index.test-d.ts:29]: Let's use a nonexistent URL instead.

```suggestion
expectType<Promise<any>>(alfy.fetch('https://foo.bar', {
```

File: package.json [hunk 19]
```diff
@@ -49,10 +50,13 @@
 		"read-pkg-up": "^6.0.0"
 	},
 	"devDependencies": {
+		"@types/got": "^9.6.12",
```
Comment (LitoMore) [inline package.json:53]: Got is written by TypeScript. Declarations are [already inside](https://github.com/sindresorhus/got/blob/f896aa52abc41fe40d4942da94a0408477358f14/package.json#L96).

**Review by LitoMore — CHANGES_REQUESTED** [review]

File: package.json (hunk 19 above)
Comment (LitoMore) [inline package.json:53]: Now this can be removed. #140 merged.

File: package.json [hunk 20]
```diff
@@ -43,7 +44,7 @@
 		"dot-prop": "^5.1.0",
 		"esm": "^3.2.18",
 		"execa": "^2.0.4",
-		"got": "^9.3.2",
+		"got": "^11.8.2",
```
Comment (LitoMore) [inline package.json:47]: These changes should not appear in this PR. Could you rebase the `main` branch?

`git rebase -i main`. Then some conflicts need to be resolved when rebasing.

Comment (jopemachine) [reply package.json:47]: If I remove `c0f6e87`, I think I also have to remove the 4 commits below to pass the tests (ava && tsd).

May I ask it is alright to remove these 5 commits?

Comment (LitoMore) [reply package.json:47]:
```shell
git checkout main
git pull upstream main
git checkout Add-types
git rebase -i main
git push origin Add-types -f
```

Comment (jopemachine) [reply package.json]: Oh, I got it, thanks for guiding me :)

**Pull request merged**